### PR TITLE
[SPARK-47329][SS][DOCS] Add note to persist dataframe while using foreachbatch and stateful streaming query to prevent state from being re-loaded in each batch

### PR DESCRIPTION
**Pull request by anishshri-db**
### What changes were proposed in this pull request?
Add note to persist dataframe while using foreachbatch and stateful streaming query to prevent state from being re-loaded in each batch

### Why are the changes needed?
Without this recommendation, if the user did not persist the dataframe in their UDF and multiple actions are called on the dataframe, the query is evaluated multiple times within the same batch leading to the state store invalidating the loaded version and reloading from cloud each time, which leads to severely degraded performance.

For eg -
Without the change, we see the same state instance loaded and committed multiple times
```
sql/core/target/unit-tests.log:596:10:17:17.090 Executor task launch worker for task 2.0 in stage 8.0 (TID 20) INFO HDFSBackedStateStoreProvider: Committed version 2 for HDFSStateStore[id=(op=0,part=2),dir=file:/Users/anish.shrigondekar/spark/spark/target/tmp/temporary-1ff51b75-034b-496d-8fc9-81df5ba78725/state/0/2] to file file:/Users/anish.shrigondekar/spark/spark/target/tmp/temporary-1ff51b75-034b-496d-8fc9-81df5ba78725/state/0/2/2.delta
sql/core/target/unit-tests.log:710:10:17:17.612 Executor task launch worker for task 2.0 in stage 11.0 (TID 26) INFO HDFSBackedStateStoreProvider: Committed version 2 for HDFSStateStore[id=(op=0,part=2),dir=file:/Users/anish.shrigondekar/spark/spark/target/tmp/temporary-1ff51b75-034b-496d-8fc9-81df5ba78725/state/0/2] to file file:/Users/anish.shrigondekar/spark/spark/target/tmp/temporary-1ff51b75-034b-496d-8fc9-81df5ba78725/state/0/2/2.delta
sql/core/target/unit-tests.log:813:10:17:18.123 Executor task launch worker for task 2.0 in stage 13.0 (TID 31) INFO HDFSBackedStateStoreProvider: Committed version 2 for HDFSStateStore[id=(op=0,part=2),dir=file:/Users/anish.shrigondekar/spark/spark/target/tmp/temporary-1ff51b75-034b-496d-8fc9-81df5ba78725/state/0/2] to file file:/Users/anish.shrigondekar/spark/spark/target/tmp/temporary-1ff51b75-034b-496d-8fc9-81df5ba78725/state/0/2/2.delta
```

With the change, we see the same state instance loaded and committed only once
```
sql/core/target/unit-tests.log:516:10:22:47.150 Executor task launch worker for task 2.0 in stage 8.0 (TID 20) INFO HDFSBackedStateStoreProvider: Committed version 2 for HDFSStateStore[id=(op=0,part=2),dir=file:/Users/anish.shrigondekar/spark/spark/target/tmp/temporary-870e4666-52dc-471e-bb19-e6c8f372d0ca/state/0/2] to file file:/Users/anish.shrigondekar/spark/spark/target/tmp/temporary-870e4666-52dc-471e-bb19-e6c8f372d0ca/state/0/2/2.delta
```

Note that we cannot always call `persist` before the `foreachbatch` function because it might lead to increased cache memory usage, possible disk writes (with the `MEMORY_AND_DISK` default level) and also trigger unwanted eviction of other critical cache blocks. Hence, we leave this responsibility to the caller.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existing unit tests

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Comment by anishshri-db** [comment]
@HeartSaVioR @sahnib @zsxwing - Could you PTAL ? Thx

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/ForeachBatchSink.scala [hunk 1]
```diff
@@ -22,19 +22,38 @@ import scala.util.control.NonFatal
...
     implicit val enc = encoder
     val ds = Dataset.ofRows(data.sparkSession, node).as[T]
-    callBatchWriter(ds, batchId)
+    // SPARK-47329 - persist the dataframe for stateful queries to prevent state stores
+    // from reloading state multiple times in each batch
+    val isStateful = isQueryStateful(data.logicalPlan)
+    if (isStateful) {
+      ds.persist()
+      callBatchWriter(ds, batchId)
+      ds.unpersist()
```
Comment (chaoqin-li1123) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/ForeachBatchSink.scala:53]: Shall we wrap this in try catch finally?

**Review by chaoqin-li1123 — COMMENTED** [review]
Good idea! I have seen people causing re-computation of stateful stages in feb without realizing it a few times.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/ForeachBatchSink.scala (hunk 1 above)
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/ForeachBatchSink.scala]: Done - I guess `try finally` should be enough here ?

Comment (chaoqin-li1123) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/ForeachBatchSink.scala]: Yes, try finally is sufficient.

**Review by Bobstar55 — APPROVED** [review]
join

**Review by Bobstar55 — APPROVED** [review]

**Review by HeartSaVioR — APPROVED** [review]
+1

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master.

**Pull request closed without merging**